new-tutorial/L4_message_passing.py

```diff
@@ -0,0 +1,386 @@
+# counterpart <3_message_passing>` are highlighted with arrows)
+# 
+
+import torch.nn as nn
+import torch.nn.functional as F
+import tqdm
+
+class SAGEConv(nn.Module):
+    """Graph convolution module used by the GraphSAGE model.
+    
+    Parameters
+    ----------
+    in_feat : int
+        Input feature size.
+    out_feat : int
+        Output feature size.
+    """
+    def __init__(self, in_feat, out_feat):
+        super(SAGEConv, self).__init__()
+        # A linear submodule for projecting the input and neighbor feature to the output.
+        self.linear = nn.Linear(in_feat * 2, out_feat)
+    
+    def forward(self, g, h):
+        """Forward computation
+        
+        Parameters
+        ----------
+        g : Graph
+            The input bipartite graph.
+        h : (Tensor, Tensor)
+            The feature of input nodes and output nodes as a pair of Tensors.
+        """
+        with g.local_scope():
+            h_src, h_dst = h
+            g.srcdata['h'] = h_src                        # <---
+            g.dstdata['h'] = h_dst                        # <---
+            # update_all is a message passing API.
+            g.update_all(fn.copy_u('h', 'm'), fn.mean('m', 'h_neigh'))
+            h_N = g.dstdata['h_N']
+            h_total = torch.cat([h_dst, h_N], dim=1)      # <---
+            return self.linear(h_total)
+
+class Model(nn.Module):
+    def __init__(self, in_feats, h_feats, num_classes):
+        super(Model, self).__init__()
+        self.conv1 = SAGEConv(in_feats, h_feats)
+        self.conv2 = SAGEConv(h_feats, num_classes)
+    
+    def forward(self, bipartites, x):
+        h_dst = x[:bipartites[0].num_dst_nodes()]
+        h = self.conv1(bipartites[0], (x, h_dst))
+        h = F.relu(h)
+        h_dst = h[:bipartites[1].num_dst_nodes()]
+        h = self.conv2(bipartites[1], (h, h_dst))
+        return h
+    
+sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+train_dataloader = dgl.dataloading.NodeDataLoader(
+    graph, train_nids, sampler,
+    batch_size=1024,
+    shuffle=True,
+    drop_last=False,
+    num_workers=0
+)
+model = Model(graph.ndata['feat'].shape[1], 128, dataset.num_classes).cuda()
+
+with tqdm.tqdm(train_dataloader) as tq:
+    for step, (input_nodes, output_nodes, bipartites) in enumerate(tq):
+        bipartites = [b.to(torch.device('cuda')) for b in bipartites]
+        inputs = node_features[input_nodes].cuda()
+        labels = node_labels[output_nodes].cuda()
+        predictions = model(bipartites, inputs)
+
+
+######################################################################
+# Both ``update_all`` and the functions in ``nn.functional`` namespace
+# support bipartite graphs, so you can migrate the code working for small
+# graphs to large graph training with minimal changes introduced above.
+# 
+
+
+######################################################################
+# Writing GNN Modules for Both Full-graph Training and Stochastic Training
```

Comment:
Universal GNN module in Three Steps ?

Reply:
"Universal" sounds too fancy.  I'd rather be specific.